### ISSUE TITLE
Problems with SemaphoreSlim on OSX

### ISSUE THREAD

**Issue by johnnyasantoss** [dotnet/corefx#42607]
Hello,
I noticed that `SemaphoreSlim.WaitAsync(int)` is sometimes extremely slow on OSX.

I was trying to set up a test and build a R2R version of a console app on Azure Pipelines, and noticed that the MacOS tests were failing on a simple wait & release [test][test1] ([logs][logs1]).

So, I created a simplified version of that test and tried to run again using the same matrix.
Here's the [repro][r1], and here is the [pipeline][p1] with the failed jobs on OSX.

BTW I'm using Linux and I don't have a machine with OSX available to test this.

This issue might be related to #28884

 [test1]: https://github.com/johnnyasantoss/ExchangeSharp/blob/e9d3589b5fa08afffd70b0fc98756f7988cee0ea/tests/ExchangeSharpTests/CryptoUtilityTests.cs#L229-L255
 [logs1]: https://dev.azure.com/johnnyasantoss/ExchangeSharp/_build/results?buildId=13&view=logs&j=2da5cbf6-3d4c-554d-b450-f46789d0dcc6&t=d93bf2ca-2f6d-5bb9-d851-d18567965641&l=23
 [r1]: https://github.com/johnnyasantoss/semaphore-slim-osx-repro
 [p1]: https://dev.azure.com/johnnyasantoss/semaphore-slim-osx-repro/_build?definitionId=2&_a=summary

**Comment by stephentoub** [comment]
(This looks buggy:
https://github.com/johnnyasantoss/semaphore-slim-osx-repro/blob/2aa38e6f27176787cc5c7f1f69cf6f4d006dd86c/tests/SemaphreRepro.Tests/UnitTest1.cs#L20-L28
You're creating the semaphore with a max count of 1.  If that Release in the thread you've started is able to execute twice before the consumer waits on the semaphore, or more generally if it gets ahead of the consumer by even a small amount, that Release call is going to throw an exception when it tries to increment the count beyond what's allowed.  That's probably unrelated, of course; if it were the issue, I'd actually expect your test to crash rather than fail due to a timeout.)

If I'm understanding your concern correctly, it's that you're having a thread sleep for 100ms between releases, and you're seeing waits on the semaphore take upwards of 172ms.  Correct?  Have you timed the actual duration of the Thread.Sleep in the instance where the wait on the semaphore takes too long?  Have you tried using something other than a SemaphoreSlim?

I suspect this isn't actually to do with SemaphoreSlim, and rather is "just" the nature of timings and thread scheduling on macOS.  Potentially related to https://github.com/dotnet/coreclr/issues/20369.

**Comment by johnnyasantoss** [comment]
> (This looks buggy:
> https://github.com/johnnyasantoss/semaphore-slim-osx-repro/blob/2aa38e6f27176787cc5c7f1f69cf6f4d006dd86c/tests/SemaphreRepro.Tests/UnitTest1.cs#L20-L28
> You're creating the semaphore with a max count of 1. If that Release in the thread you've started is able to execute twice before the consumer waits on the semaphore, or more generally if it gets ahead of the consumer by even a small amount, that Release call is going to throw an exception when it tries to increment the count beyond what's allowed. That's probably unrelated, of course; if it were the issue, I'd actually expect your test to crash rather than fail due to a timeout.)

Yep its buggy, but it is just to demonstrate the delay between the release and the actual release of the semaphore.

> If I'm understanding your concern correctly, it's that you're having a thread sleep for 100ms between releases, and you're seeing waits on the semaphore take upwards of 172ms. Correct?

Yes

> Have you timed the actual duration of the Thread.Sleep in the instance where the wait on the semaphore takes too long? Have you tried using something other than a SemaphoreSlim?

Not yet, but will do.

> I suspect this isn't actually to do with SemaphoreSlim, and rather is "just" the nature of timings and thread scheduling on macOS. Potentially related to dotnet/coreclr#20369.

Thanks, I'll investigate this further and time the `Thread.Sleep`s in the tests.

**Comment by johnnyasantoss** [comment]
Apparently you were right. The `Thread.Sleep` calls are taking a lot more time than it should. (see [this log](https://dev.azure.com/johnnyasantoss/semaphore-slim-osx-repro/_build/results?buildId=50&view=logs&j=e44bd064-d8d9-5feb-6343-09b8c49ab1a8&t=85a1ad8e-4daf-5689-e663-871501e587ee&l=48))
